nilearn/signal.py

```diff
@@ -452,8 +453,22 @@ def clean(signals, detrend=True, standardize=True, confounds=None,
 
         confounds = _ensure_float(confounds)
         confounds = _standardize(confounds, normalize=True, detrend=detrend)
-        Q = linalg.qr(confounds, mode='economic')[0]
-        signals -= np.dot(Q, np.dot(Q.T, signals))
+
+        if (LooseVersion(scipy.__version__) > LooseVersion('0.9.0')):
+            # Pivoting in qr decomposition was added in scipy 0.10
+            Q, R, _ = linalg.qr(confounds, mode='economic', pivoting=True)
+            Q = Q[:, np.abs(np.diag(R)) > np.finfo(np.float).eps * 100.]
+            signals -= Q.dot(Q.T).dot(signals)
+        else:
+            Q, R = linalg.qr(confounds, mode='economic')
+            non_null_diag = np.abs(np.diag(R)) > np.finfo(np.float).eps * 100.
+            if np.all(non_null_diag):
+                signals -= Q.dot(Q.T).dot(signals)
+            elif np.any(non_null_diag):
+                R = R[:, non_null_diag]
+                confounds = confounds[:, non_null_diag]
+                inv = scipy.linalg.inv(np.dot(R.T, R))
+                signals -= confounds.dot(inv).dot(confounds.T).dot(signals)
```

Comment:
How about we cut our losses here and stop discussing minute details to
focus on the big picture: the problem is solved, and we want to release
0.1.3 to get bug fixes out to our users.

You may perfectly be right. I would welcome a PR with corresponding
timings showing that it is possible to simplify the code with the same
numerical stability and speed. But in the mean time, I really want to get
0.1.3 out.